XD-3381: Provide test support for modules

ericbottard opened the pull request:
Example usage is here: https://github.com/ericbottard/spring-cloud-stream-modules/commit/3ad04862f8f844e23c51b09a52c5adfb58000d10

dturanski commented on pom.xml:
```diff
@@ -34,7 +34,8 @@
 		<module>spring-cloud-stream-samples</module>
 		<module>docs</module>
-	</modules>
+    <module>spring-cloud-stream-test-support</module>
```
indent

dturanski commented on pom.xml:
```diff
@@ -147,4 +148,4 @@
 		</pluginRepository>
 	</pluginRepositories>
-</project>
+</project>
```
new line

dturanski commented on spring-cloud-stream-test-support/pom.xml:
```diff
@@ -0,0 +1,43 @@
+	<groupId>org.springframework.cloud</groupId>
+	<artifactId>spring-cloud-stream-test-support</artifactId>
+	<name>spring-cloud-stream-test-support</name>
+	<url>http://maven.apache.org</url>
```
remove url

dturanski commented on spring-cloud-stream-test-support/pom.xml:
```diff
@@ -0,0 +1,43 @@
+		<dependency>
+			<groupId>org.springframework.cloud</groupId>
+			<artifactId>spring-cloud-stream-codec</artifactId>
+		</dependency>
```
shouldn't need codec

dturanski commented on spring-cloud-stream-test-support/src/main/java/org/springframework/cloud/stream/test/binder/TestSupportBinder.java:
```diff
@@ -0,0 +1,103 @@
+	@Override
+	public void bindProducer(String name, MessageChannel outboundBindTarget, Properties properties) {
+		BridgeHandler bridge = new BridgeHandler();
+		QueueChannel queue = new QueueChannel();
```
not sure why we need this to be asynch

ericbottard replied:
I'm afraid I don't understand what you mean. I think, on the contrary, that we prefer to be synchronous/blocking when writing tests.

dturanski replied:
Why not just 
   
       (SubscribableChannel)outputBindTarget).subscribe(...)

ericbottard replied:
Good point. At first, I thought the `MessageChannelBinderSupport.Binding` concept would help me, and thus I stuck with it. Not sure how much like a "real" binder we want this to feel

dturanski replied:
Yeah. I'm not sure at this point.

dturanski commented on spring-cloud-stream-test-support/src/main/java/org/springframework/cloud/stream/test/binder/TestSupportBinderAutoConfiguration.java:
```diff
@@ -0,0 +1,42 @@
+ */
+@Configuration
+@ConditionalOnMissingBean(Binder.class)
+@AutoConfigureOrder(Ordered.LOWEST_PRECEDENCE - 10)
```
I think this should be HIGHEST_PRECEDENCE since we don't ever want another Binder implementation if this is on the classpath. and no ConditionalOnMissingBean

dturanski replied:
I did it a bit different here:
@Configuration
@Order(Ordered.HIGHEST_PRECEDENCE)
public class TestSupportBinderAutoConfiguration {

	@Bean
	public MessageCache messageCache() {
		return new MessageCache();
	}

	@Bean
	public Binder testSupportBinder() {
		return new TestSupportBinder(messageCache());
	}
}

So tests can @Autowire the MessageCache and use that instead of the Binder

ericbottard replied:
Interesting. I assume the MessageCache has a mapping between Channels and .. something?

But then, how do you use it?

ericbottard replied:
I guess one could do
```
assertThat(messageCache.for(Processor.output()), receivedMessageThat(...))
```

ericbottard replied:
A bit more verbose, but less static. I like it. Not a big fan of the "cache" name, though

mbogoevici replied:
+1

ericbottard replied:
Should we go the extra mile and provide support for injecting the message results _per channel_ (given that most modules will have only one output channel), eg
```
@Autowired
@ResultsFor(Processor.OUTPUT)
private Queue<Message<?>> result;

...

assertThat(result, receivesPayloadThat(equalTo("hellofoo")).within(10));
```

dturanski replied:
I think so

dturanski replied:
How about  s/MessageCache/TestMessageConsumer ?

dturanski commented:
Need to add some tests

dturanski commented on spring-cloud-stream-test-support/src/main/resources/META-INF/spring.factories:
```diff
@@ -0,0 +1,2 @@
+org.springframework.boot.autoconfigure.EnableAutoConfiguration:\
+org.springframework.cloud.stream.test.binder.TestSupportBinderAutoConfiguration
```
new line

mbogoevici commented on spring-cloud-stream-test-support/src/main/java/org/springframework/cloud/stream/test/binder/TestSupportBinder.java:
```diff
@@ -0,0 +1,103 @@
+ * @author Eric Bottard
+ * @see org.springframework.cloud.stream.test.matcher.MessageChannelMatcher
+ */
+public class TestSupportBinder extends MessageChannelBinderSupport {
```
I would call this `TestBinder` (it's easier to read)

dturanski replied:
Yes but we have https://github.com/spring-cloud/spring-cloud-stream/blob/master/spring-cloud-stream-binders/spring-cloud-stream-binder-test/src/main/java/org/springframework/cloud/stream/binder/AbstractTestBinder.java  so may be confusing

mbogoevici commented on spring-cloud-stream-test-support/src/main/java/org/springframework/cloud/stream/test/binder/TestSupportBinder.java:
```diff
@@ -0,0 +1,103 @@
+ */
+public class TestSupportBinder extends MessageChannelBinderSupport {
+
+	public static Map<MessageChannel, QueueChannel> channelToBindings =
```
Perhaps this doesn't need to be static, and could be injectable as a bean - thus saving the need to reference directly the binder and enforcing isolation between tests?

ericbottard replied:
It all comes down to how it reads in _e.g._ a Hamcrest matcher (see discussion below).
At first, I really wanted to be able to say `assertThat(theChannel, ...)`

mbogoevici replied:
Right, I think that the discussion below provides a very good alternative :)

ericbottard commented:
Updated the PR to use injection rather than static access.
Added extensive tests.

Yet, I have not gone through my proposed idea to inject _per-channel_, because I think it is hard to read in tests. I think it is important to see 
```
doSomethingWith(input())
// and just after
assertThat(someReferenceTo(output()), ...)
```

Annotating with _e.g._ `Processor.OUTPUT` has two drawbacks IMO:
* the reference is at the injection point, not the assertion point
* it is not obvious that `Processor.OUTPUT` relates to `output()`

WDYT?

markfisher commented:
Is it worth considering a Gateway as the component that could be injected into test code? That would provide `sendAndReceive` from the perspective of the test *as a client*.

ericbottard commented:
@markfisher could you elaborate (if I understand what you're saying, I believe it only makes sense for Processor, correct?). And would not cover cases like 1 input for 2 outputs, _etc._

markfisher commented:
actually the spring-integration Gateway supports: send, receive, and sendAndReceive, so it could be generally useful as the test "client" cc @garyrussell

ericbottard commented:
Still not sure I get it. Could you provide an example usage (from the perspective of test writer) of say, testing a routing module that has one input and 2 outputs ("odd" and "even"):
```
public interface MyChannels {
   @Input("input")
   MessageChannel input();

   @Ouput("odd")
   MessageChannel odd();

   @Output("even")
   MessageChannel even();
}
```

BTW, aren't the types in Sink & Source currently backwards (assuming we want/need the down-typing to `SubscribableChannel`)

markfisher commented:
They are not backwards. The "direction" of those channels is from the perspective of the module. That is why I was suggesting looking at the test from a "client" perspective (outside of the black-box). The test code would send to the module's input, and receive from the module's output.

mbogoevici commented:
@ericbottard Personally, I think that relying on the interface methods - i.e.  `input()` vs INPUT is much better, too. Personally, I wouldn't register channel beans at all, weren't for a total pain to use things like `@ServiceActivator` that rely on finding channels by name. 

I would agree that we could provide a test Gateway, but can we do it as a future enhancement ? Also, can't see how we can use it for single-channel modules (e.g. sources and sinks). I think it's a valuable idea, but not having support is not a blocker for solving the immediate problem of having a test binder?

markfisher commented:
one thing I was discussing with @markpollack is that we could inject the "other side" into the test... for example if we are testing a Source module (which sends messages to its own "output" channel), the test itself could be injected with an `@Input`. Essentially the test is acting as if it's a direct-bound Sink in that case (and vice versa of course).

mbogoevici commented:
@markfisher that is taking things in a completely different direction, and I am afraid it both complicates things and makes them very spring-cloud-data centric. In the current approach, a tester will get access to all the 'bound' interfaces of a module (which they know very well) and can act and test them directly. In the approach where we bind and bridge, they would have to deal with an additional set of channels, and it is not very clear where are they coming from. Additionally, for multi-inputs/multi-outputs, you'd have to be fairly descriptive about the binding.

markpollack commented:
For source, it would seem a more traditional approach of the test executing the stimulus the source expects, http,jms message, rabbit etc, would be very straightforward.  See https://github.com/spring-projects/spring-xd/blob/master/extensions/spring-xd-extension-http/src/test/java/org/springframework/integration/x/http/NettyHttpInboundChannelAdapterTests.java

I like the Gateway approach for processors.

markfisher commented on pom.xml:
```diff
@@ -34,7 +34,8 @@
 		<module>docs</module>
-	</modules>
+		<module>spring-cloud-stream-test-support</module>
+  </modules>
```
fix indentation

markfisher commented on spring-cloud-stream-test-support/src/main/java/org/springframework/cloud/stream/test/binder/MessageCollector.java:
```diff
@@ -0,0 +1,50 @@
+		Assert.notNull(results.remove(channel));
+	}
+
+	public BlockingQueue<Message<?>> forChannel(MessageChannel channel) {
```
should this call register if not already registered?

ericbottard replied:
Same reason as above, we may even want to fail if not found, as this would happen from a misuse of the test support IMO

markfisher replied:
also this needs javadoc

markfisher replied:
it should either fail or register on-demand (latter might be better), but returning null here seems dangerous

markfisher commented on spring-cloud-stream-test-support/src/main/java/org/springframework/cloud/stream/test/binder/MessageCollector.java:
```diff
@@ -0,0 +1,50 @@
+
+	/*package*/ BlockingQueue register(MessageChannel channel) {
+		LinkedBlockingDeque<Message<?>> result = new LinkedBlockingDeque<>();
+		Assert.isNull(results.put(channel, result));
```
this assertion is too late - the new empty queue has already been *put* into the map, but that is what needs to be prevented (also the Assert should provide a useful message).

ericbottard replied:
`put` returns what was there _before_. I want to fail if there was something already registered.

markfisher replied:
What I was saying is that you are not *preventing* the change in the map this way. Sure, the assert fails, but the damage has already been done.

ericbottard replied:
Yes, I know. But as you say, there is will be an exception just after. Can refactor so that it does not mutate.

markfisher replied:
My point was that a caller could handle the exception and not be stuck with an invalid MessageCollector, whereas the current code is making that MessageCollector invalid regardless of how the exception is handled.

You can use `putIfAbsent` so otherwise it would be the same (although that is only added to the core Map in 1.8, so would need to use ConcurrentMap, which would be fine).

ericbottard replied:
This class should really only be used by the TestSupportBinder (which you
can tell from the two package protected methods).
We may even want to extract an interface for the `forChannel()` method and
move this impl. to be private to the binder

On Fri, Aug 14, 2015 at 4:43 PM, Mark Fisher <notifications@github.com>
wrote:

> In
> spring-cloud-stream-test-support/src/main/java/org/springframework/cloud/stream/test/binder/MessageCollector.java
> <https://github.com/spring-cloud/spring-cloud-stream/pull/72#discussion_r37082773>
> :
>
> > +import org.springframework.messaging.MessageChannel;
> > +import org.springframework.util.Assert;
> > +
> > +/**
> > + * Maintains a map between (output) channels and messages received (in FIFO order). To be injected in tests that
> > + * can then run assertions on the enqueued messages.
> > + *
> > + * @author Eric Bottard
> > + */
> > +public class MessageCollector {
> > +
> > +	private Map<MessageChannel, BlockingQueue<Message<?>>> results = new HashMap<>();
> > +
> > +	/*package*/ BlockingQueue register(MessageChannel channel) {
> > +		LinkedBlockingDeque<Message<?>> result = new LinkedBlockingDeque<>();
> > +		Assert.isNull(results.put(channel, result));
>
> My point was that a caller could handle the exception and not be stuck
> with an invalid MessageCollector, whereas the current code is making that
> MessageCollector invalid regardless of how the exception is handled.
>
> You can use putIfAbsent so otherwise it would be the same (although that
> is only added to the core Map in 1.8, so would need to use ConcurrentMap,
> which would be fine).
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/spring-cloud/spring-cloud-stream/pull/72/files#r37082773>
> .
>

markfisher replied:
If that's really true, can it be encapsulated completely into the binder? Even so, I don't see why that precludes making the assertion proactive instead of reactive etc. That makes the code easier to understand if nothing else.

markfisher commented on spring-cloud-stream-test-support/src/main/java/org/springframework/cloud/stream/test/binder/MessageCollector.java:
```diff
@@ -0,0 +1,50 @@
+	}
+
+	/*package*/ void unregister(MessageChannel channel) {
+		Assert.notNull(results.remove(channel));
```
does this really need to Assert? maybe return a boolean instead, but I don't see why it can't be considered harmless to unregister something that does not exist

ericbottard replied:
My reasoning is that this is being called from the TestSupportBus only, it should never try to unregister something it did not register itself in the first place

markfisher replied:
JavaDoc should be added including a `@throws` description (same for `register`), but I still don't see why an exception needs to be thrown in the case of unregistering a channel that was not registered.

markfisher commented on spring-cloud-stream-test-support/src/main/java/org/springframework/cloud/stream/test/binder/TestSupportBinder.java:
```diff
@@ -0,0 +1,129 @@
+
+	@Override
+	public void bindConsumer(String name, MessageChannel inboundBindTarget, Properties properties) {
+		// Do nothing. A module author can grab hold forChannel the input channel
```
comment grammar (should it be at javadoc level?)

ericbottard replied:
Hmm, weird sentence indeed. May be IDEA refactoring from text in comments. The idea is already expressed in the class javadoc anyhow. Will remove

markfisher commented on spring-cloud-stream-test-support/src/main/java/org/springframework/cloud/stream/test/matcher/MessageQueueMatcher.java:
```diff
@@ -0,0 +1,160 @@
+/**
+ * A Hamcrest Matcher meant to be used in conjunction with {@link TestSupportBinder}.
+ *
+ * <p>Expected usage is forChannel the form (with appropriate static imports):
```
this doesn't make sense

ericbottard replied:
Hmm, same "forChannel" occurence. Damn IDEA. Will revisit

ericbottard commented on spring-cloud-stream-test-support/src/main/java/org/springframework/cloud/stream/test/binder/TestSupportBinder.java:
```diff
@@ -0,0 +1,129 @@
+	}
+
+	/**
+	 * Bridges a QueueChannel to the module output channel, so that it can
```
Outdated javadoc, will remove

markfisher commented on spring-cloud-stream-test-support/src/main/java/org/springframework/cloud/stream/test/matcher/MessageQueueMatcher.java:
```diff
@@ -0,0 +1,160 @@
+ *
+ * @author Eric Bottard
+ */
+public class MessageQueueMatcher<T> extends BaseMatcher<BlockingQueue<Message<?>>> {
```
This class has a lot of warnings, such as "rawtypes" (showing up in the IDE). Also, is there any way to avoid some of the "unchecked" suppressions via parameterization?

ericbottard commented:
Rebased, force pushed

ericbottard commented:
Bumped. Would be nice to have some closure on this (I can merge myself if everyone is ok with its current state)

dturanski commented:
squashed and merged as e2497ef
